src/pe.rs

```diff
@@ -0,0 +1,2984 @@
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+pub struct ImageRomOptionalHeader {
+    pub magic: U16<LE>,
+    pub major_linker_version: u8,
+    pub minor_linker_version: u8,
+    pub size_of_code: U32<LE>,
+    pub size_of_initialized_data: U32<LE>,
+    pub size_of_uninitialized_data: U32<LE>,
+    pub address_of_entry_point: U32<LE>,
+    pub base_of_code: U32<LE>,
+    pub base_of_data: U32<LE>,
+    pub base_of_bss: U32<LE>,
+    pub gpr_mask: U32<LE>,
+    pub cpr_mask: [U32<LE>; 4],
+    pub gp_value: U32<LE>,
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+pub struct ImageOptionalHeader64 {
+    pub magic: U16<LE>,
+    pub major_linker_version: u8,
+    pub minor_linker_version: u8,
+    pub size_of_code: U32<LE>,
+    pub size_of_initialized_data: U32<LE>,
+    pub size_of_uninitialized_data: U32<LE>,
+    pub address_of_entry_point: U32<LE>,
+    pub base_of_code: U32<LE>,
+    pub image_base: U64<LE>,
+    pub section_alignment: U32<LE>,
+    pub file_alignment: U32<LE>,
+    pub major_operating_system_version: U16<LE>,
+    pub minor_operating_system_version: U16<LE>,
+    pub major_image_version: U16<LE>,
+    pub minor_image_version: U16<LE>,
+    pub major_subsystem_version: U16<LE>,
+    pub minor_subsystem_version: U16<LE>,
+    pub win32_version_value: U32<LE>,
+    pub size_of_image: U32<LE>,
+    pub size_of_headers: U32<LE>,
+    pub check_sum: U32<LE>,
+    pub subsystem: U16<LE>,
+    pub dll_characteristics: U16<LE>,
+    pub size_of_stack_reserve: U64<LE>,
+    pub size_of_stack_commit: U64<LE>,
+    pub size_of_heap_reserve: U64<LE>,
+    pub size_of_heap_commit: U64<LE>,
+    pub loader_flags: U32<LE>,
+    pub number_of_rva_and_sizes: U32<LE>,
+    //pub data_directory: [ImageDataDirectory; IMAGE_NUMBEROF_DIRECTORY_ENTRIES],
+}
+
+pub const IMAGE_NT_OPTIONAL_HDR32_MAGIC: u16 = 0x10b;
+pub const IMAGE_NT_OPTIONAL_HDR64_MAGIC: u16 = 0x20b;
+pub const IMAGE_ROM_OPTIONAL_HDR_MAGIC: u16 = 0x107;
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+pub struct ImageNtHeaders64 {
+    pub signature: U32<LE>,
+    pub file_header: ImageFileHeader,
+    pub optional_header: ImageOptionalHeader64,
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+pub struct ImageNtHeaders32 {
+    pub signature: U32<LE>,
+    pub file_header: ImageFileHeader,
+    pub optional_header: ImageOptionalHeader32,
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+pub struct ImageRomHeaders {
+    pub file_header: ImageFileHeader,
+    pub optional_header: ImageRomOptionalHeader,
+}
+
+// IMAGE_FIRST_SECTION doesn't need 32/64 versions since the file header is the same either way.
+
+/*
+#define IMAGE_FIRST_SECTION( ntheader ) ((PIMAGE_SECTION_HEADER)        \
+    ((ULONG_PTR)(ntheader) +                                            \
+     FIELD_OFFSET( IMAGE_NT_HEADERS, OptionalHeader ) +                 \
+     ((ntheader))->FileHeader.SizeOfOptionalHeader   \
+    ))
+    */
```

Comment:
Should this be deleted?

Reply:
Yep, I kept this for reference initially, but the parser does the equivalent of this now.